rotkehlchen/chain/ethereum/modules/cowswap/decoder.py

```diff
@@ -0,0 +1,236 @@
+import logging
+from typing import TYPE_CHECKING, Any, Callable, NamedTuple, Optional
+
+from rotkehlchen.accounting.structures.balance import Balance
+from rotkehlchen.accounting.structures.base import HistoryBaseEntry
+from rotkehlchen.accounting.structures.types import HistoryEventSubType, HistoryEventType
+from rotkehlchen.assets.asset import Asset, EvmToken
+from rotkehlchen.chain.ethereum.modules.cowswap.constants import CPT_COWSWAP
+from rotkehlchen.chain.ethereum.utils import asset_normalized_value
+from rotkehlchen.chain.evm.constants import ETH_SPECIAL_ADDRESS
+from rotkehlchen.chain.evm.decoding.interfaces import DecoderInterface
+from rotkehlchen.chain.evm.decoding.structures import ActionItem
+from rotkehlchen.chain.evm.structures import EvmTxReceiptLog
+from rotkehlchen.chain.evm.types import string_to_evm_address
+from rotkehlchen.constants.assets import A_ETH
+from rotkehlchen.constants.resolver import evm_address_to_identifier
+from rotkehlchen.fval import FVal
+from rotkehlchen.logging import RotkehlchenLogsAdapter
+from rotkehlchen.types import ChainID, ChecksumEvmAddress, EvmTokenKind, EvmTransaction
+from rotkehlchen.utils.misc import hex_or_bytes_to_address, hex_or_bytes_to_int
+
+
+if TYPE_CHECKING:
+    from rotkehlchen.assets.asset import CryptoAsset
+    from rotkehlchen.chain.ethereum.node_inquirer import EthereumInquirer
+    from rotkehlchen.chain.evm.decoding.base import BaseDecoderTools
+    from rotkehlchen.user_messages import MessagesAggregator
+
+logger = logging.getLogger(__name__)
+log = RotkehlchenLogsAdapter(logger)
+
+TRADE_SIGNATURE = b'\xa0zT:\xb8\xa0\x18\x19\x8e\x99\xca\x01\x84\xc9?\xe9\x05\ny@\n\nr4A\xf8M\xe1\xd9r\xcc\x17'  # noqa: E501
+PLACE_ETH_ORDER_SIGNATURE = b"\xcf_\x9d\xe2\x98A2&R\x03\xb5\xc35\xb2W\'p,\xa7rb\xffb.\x13k\xaasb\xbf\x1d\xa9"  # noqa: E501
+REFUND_ETH_ORDER_SIGNATURE = b'\x19Rq\x06\x8a(\x81\x91\xe4\xb2e\xc6A\xa5k\x982\x91\x9fi\xe9\xe7\xd6\xc2\xf3\x1b\xa4\x02x\xae\xb8Z'  # noqa: E501
+INVALIDATE_ETH_ORDER_SIGNATURE = b'\xb8\xba\xd1\x02\xac\x8b\xba\xcf\xef1\xff\x1c\x90n\xc6\xd9Q\xc20\xb4\xdc\xe7P\xbb\x03v\xb8\x12\xad5\x85*'  # noqa: E501
+
+GPV2_SETTLEMENT_ADDRESS = string_to_evm_address('0x9008D19f58AAbD9eD0D60971565AA8510560ab41')
+ETH_FLOW_ADDRESS = string_to_evm_address('0x40A50cf069e992AA4536211B23F286eF88752187')
+
+
+class TradeData(NamedTuple):
```

Comment:
Yeah, is a generic structure indeed. I'll think on a place where to move it.

Sub-swaps thing was discussed, but not implemented (since it is just something "nice to have" and does not have a high priority).